File: machine/ts/src/machine/devices/audio/selected_slot_latch.ts
import {
	APU_PARAMETER_SOURCE_ADDR_INDEX,
	APU_SLOT_COUNT,
	APU_STATUS_SELECTED_SLOT_ACTIVE,
} from './contracts';
import type { ApuSlotBank } from './slot_bank';
import { IO_APU_SELECTED_SOURCE_ADDR, IO_APU_SLOT } from '../../bus/io';
import type { Memory } from '../../memory/memory';
import type { DeviceStatusLatch } from '../device_status';

export class ApuSelectedSlotLatch {
	public constructor(
		private readonly memory: Memory,
		private readonly status: DeviceStatusLatch,
		private readonly slots: ApuSlotBank,
	) {}

	public reset(): void {
		this.memory.writeValue(IO_APU_SELECTED_SOURCE_ADDR, 0);
		this.status.setStatusFlag(APU_STATUS_SELECTED_SLOT_ACTIVE, false);
	}

	public refresh(): void {
		const slot = this.memory.readIoU32(IO_APU_SLOT);
		const active = slot < APU_SLOT_COUNT && (this.slots.activeMask & (1 << slot)) !== 0;
		this.memory.writeIoValue(IO_APU_SELECTED_SOURCE_ADDR, active ? this.slots.registerWord(slot, APU_PARAMETER_SOURCE_ADDR_INDEX) : 0);
		this.status.setStatusFlag(APU_STATUS_SELECTED_SLOT_ACTIVE, active);
	}
}

File: machine/cpp/machine/devices/audio/selected_slot_latch.cpp
#include "machine/devices/audio/selected_slot_latch.h"

#include "spec/bmsx/io.h"
#include "spec/audio/apu.h"
#include "machine/devices/audio/slot_bank.h"
#include "machine/devices/device_status.h"
#include "machine/memory/memory.h"

namespace bmsx {

ApuSelectedSlotLatch::ApuSelectedSlotLatch(Memory& memory, DeviceStatusLatch& status, ApuSlotBank& slots)
	: m_memory(memory)
	, m_status(status)
	, m_slots(slots) {}

void ApuSelectedSlotLatch::refresh(ApuAudioSlot slot) {
	const bool active = (m_slots.activeMask() & (1u << slot)) != 0u;
	m_memory.writeIoU32(IO_APU_SELECTED_SOURCE_ADDR, active ? m_slots.registerWord(slot, APU_PARAMETER_SOURCE_ADDR_INDEX) : 0u);
	m_status.setStatusFlag(APU_STATUS_SELECTED_SLOT_ACTIVE, active);
}

} // namespace bmsx
